---
title: "Susceptibility data"
output:
  html_document:
    keep_md: yes
    css: "css/main.css"
---

Report generated at: `r format(as.POSIXct(Sys.time(), format = "%y%m%d %H:%M"), "%m/%d/%Y %I:%M %p")` PDT

```{r setup, include=FALSE}
library(dplyr)
library(ggplot2)
library(gridExtra)
knitr::opts_chunk$set(echo = FALSE)
source("./lib/readers.R", chdir = TRUE)
source("./lib/datatables.R", chdir = TRUE)

ControlVariantNames = c(
  "Control",
  "Wuhan",
  "S:614G"
)
CPInfectionNames = c(
  "S:614G",
  "",
  "Wuhan"
)

variantSynonyms = c(
  "B.1.1.7 full genome" = "B.1.1.7 Spike",
  "B.1.351 full genome" = "B.1.351 Spike",
  "P.1 full genome" = "P.1 Spike",
  "B.1.429 full genome" = "B.1.429 Spike",
  "B.1.427 full genome" = "B.1.429 Spike",
  "S:501Y+614G" = "S:501Y",
  "S:484K+614G" = "S:484K",
  "S:452R+614G" = "S:452R",
  "S:417N+614G" = "S:417N",
  "S:439K+614G" = "S:419K",
  "S:453F+614G" = "S:453F"
)
partialResistFold = 3
resistFold = 10

dfSuscResultsMAb = read.suscResultsMAb()
dfSuscResultsMAb = dfSuscResultsMAb %>%
  filter(control_variant_name %in% ControlVariantNames) %>%
  mutate(
    variant_name = ifelse(
      variant_name %in% names(variantSynonyms),
      yes = variantSynonyms[variant_name],
      no = variant_name
    )
  )
dfSuscResultsCP = read.suscResultsCP(
  partialResistFold = partialResistFold,
  resistFold = resistFold
)
dfSuscResultsCP = dfSuscResultsCP %>%
  filter(control_variant_name %in% ControlVariantNames) %>%
  filter(infection %in% CPInfectionNames)

dfSuscResultsIP = read.suscResultsIP(
  partialResistFold = partialResistFold,
  resistFold = resistFold
)
dfSuscResultsIP = dfSuscResultsIP %>%
  filter(control_variant_name %in% ControlVariantNames)

dfSuscResultsCP$group = "CP"
dfSuscResultsIP$group = "IP"
dfSuscResultsPlasma = bind_rows(dfSuscResultsCP, dfSuscResultsIP)
dfSuscResultsPlasma = dfSuscResultsPlasma %>%
  mutate(
    variant_name = ifelse(
      variant_name %in% names(variantSynonyms),
      yes = variantSynonyms[variant_name],
      no = variant_name
    )
  )

dfVirusVariants = read.virusVariants() %>%
  select(variant_name, num_muts)
dfAntibodies = read.antibodies()





```


## Statistic Numbers

```{r numbers}
inTrialMAbs = dfAntibodies %>%
  filter(!is.na(availability))

structuredMAbs = dfAntibodies %>%
  filter(!is.na(pdb_id))

dfSuscResults = dfSuscResultsMAb %>%
  mutate(
    group = "MAb",
    rx_group = ifelse(
      Vectorize(function(x) length(setdiff(x, inTrialMAbs$ab_name)) == 0)(ab_name),
      yes = "phase3MAb",
      no = ifelse(
        Vectorize(function(x) length(setdiff(x, structuredMAbs$ab_name)) == 0)(ab_name),
        yes = "structuredMAb",
        no = "otherMAb"
      )
    )
  ) %>%
  bind_rows(dfSuscResultsPlasma) %>%
  merge(dfVirusVariants, by = "variant_name") %>%
  mutate(
    rx_group = factor(
      ifelse(
        group %in% c("CP", "IP"),
        yes = group,
        no = rx_group
      ),
      levels = c("CP", "IP", "phase3MAb", "structuredMAb", "otherMAb")
    ),
    variant_group = factor(
      ifelse(
        variant_name %in% c("S:501Y", "S:484K", "B.1.1.7 Spike", "B.1.351 Spike", "P.1 Spike"),
        yes = variant_name,
        no = ifelse(
          num_muts > 1,
          yes = "Other mutation combinations",
          no = "Other mutations"
        )
      ),
      levels = c("S:501Y", "S:484K", "Other mutations", "B.1.1.7 Spike",
                 "B.1.351 Spike", "P.1 Spike", "Other mutation combinations")
    )
  )

aggregate(cbind(num = ref_name) ~ rx_group + variant_group,
          dfSuscResults,
          FUN = length) %>%
  reshape(idvar = "variant_group", timevar = "rx_group", direction = "wide") %>%
  styledKBL

```

### Unique MAbs

```{r unique-mabs}
dfUsedMAbs = dfAntibodies %>%
  filter(ab_name %in% unique(unlist(dfSuscResultsMAb$ab_name))) %>%
  mutate(has_structure = factor(ifelse(is.na(pdb_id), "No", "Yes"), levels = c("Yes", "No")))

aggregate(cbind(count = ab_name) ~ has_structure, dfUsedMAbs, FUN = length) %>%
  styledKBL
```

## Convalescent and immune plasma

<span style="padding: .1rem .2rem; margin: 0 .2rem; border: 1px solid black; background-color: #146aa8; color: white;">fold ↓ susceptibility > `r resistFold`</span>&nbsp;
<span style="padding: .1rem .2rem; margin: 0 .2rem; border: 1px solid black; background-color: #7fcbee; color: black;">fold ↓ susceptibility `r partialResistFold` to `r resistFold`</span>&nbsp;
<span style="padding: .1rem .2rem; margin: 0 .2rem; border: 1px solid black; background-color: #ffffff; color: black;">fold ↓ susceptibility < `r partialResistFold`</span>

```{r plasma-fold-fig, fig.height = 10, fig.width = 16, fig.retina = 4, message = FALSE}

groupTitle = c(
  "CP" = "Convalescent plasma",
  "IP" = "Vaccinated plasma"
)

paddingTitle <- function(title) {
  paste("  ", title)
}

variantTitle = c(
  "B.1.1.7 Spike" = "B.1.1.7",
  "B.1.351 Spike" = "B.1.351",
  "P.1 Spike" = "P.1",
  # "S:452R" = "L452R",
  "B.1.429 Spike" = "B.1.427/9",
  "S:501Y" = "N501Y",
  "S:484K" = "E484K"
  #"S:417N" = "417N",
  #"S:439K" = "439K",
  #"S:453F" = "453F"
)
filterVariants = names(variantTitle)
excludedRefRxNames = c(
  "Planas21 CP_M3",
  "Planas21 CP_M6",
  "Planas21 BNT_W2",
  "Planas21 BNT_W3",
  "Planas21 BNT_W4",
  "Planas21 BNT_W5",
  "Garcia-Beltran21 BNT",
  "Garcia-Beltran21 Mod",
  "Widera21 CP LiveVirus",
  "Widera21 BNT LiveVirus",
  "Brown21 CP",
  "Brown21 BNT",
  "Supasa21 CP LiveVirus",
  "Supasa21 BNT LiveVirus",
  "Supasa21 AZD LiveVirus",
  "Edara21 CP_acute LiveVirus",
  "Kuzmina21 CP",
  "Kuzmina21 BNT",
  "Skelly21 BNT",
  "Skelly21 BNT_sd LiveVirus",
  "Zuckerman21 BNT LiveVirus",
  "McCallum21b BNT_antibody",
  "McCallum21b BNT_infected"
)
shortRefRxNames = c(
  "Becker21 CP LiveVirus" = "Becker21*\n",
  "Becker21 BNT LiveVirus" = "BNT\nBecker21*",
  "Brown21 CP" = "Brown21\n",
  "Brown21 CP LiveVirus" = "Brown21*\n",
  "Brown21 BNT" = "BNT\nBrown21",
  "Brown21 BNT LiveVirus" = "BNT\nBrown21*",
  "Collier21 BNT" = "BNT\nCollier21",
  "Collier21 BNT_1ds" = "BNT_1ds\nCollier21",
  "Collier21 CP" = "Collier21\n",
  "Deng21 CP LiveVirus" = "Deng21*\n",
  "Deng21 BNT LiveVirus" = "BNT\nDeng21*",
  "Deng21 Mod LiveVirus" = "Mod\nDeng21*",
  "Chen21 CP LiveVirus" = "Chen21*",
  "Chen21 BNT LiveVirus" = "BNT\nChen21*",
  "Emary21 AZD LiveVirus" = "AZD\nEmary21",
  "Edara21 CP LiveVirus" = "Edara21*\n",
  "Edara21 Mod LiveVirus" = "Mod\nEdara21*",
  "Faulkner21 CP_Mild LiveVirus" = "Mild\nFaulkner21*",
  "Faulkner21 CP_Patient LiveVirus" = "Severe\nFaulkner21*",
  "Greaney21 CP" = "Greaney21\n",
  "Hoffmann21 BNT" = "BNT\nHoffmann21",
  "Hoffmann21 CP" = "Hoffmann21\n",
  "Hu21 1M" = "1M\nHu21",
  "Hu21 8M" = "8M\nHu21",
  "Jangra21 CP" = "Jangra21\n",
  "Jangra21 BNT" = "BNT\nJangra21",
  "Liu21 CP" = "Liu21\n",
  "Liu21b BNT" = "BNT\nLiu21b",
  "Muik21 BNT" = "BNT\nMuik21",
  "Moyo-Gwete21 CP" = "Moyo-G21\n",
  "Madhi21 AZD" = "AZD\nMadhi21",
  "Marot21 CP LiveVirus" = "Marot21*\n",
  "Marot21 BNT LiveVirus" = "BNT\nMarot21*",
  "Planas21 BNT_W4" = "BNT\nPlanas21",
  "Rees-Spear21 mild" = "Mild\nRSC21",
  "Rees-Spear21 severe" = "Severe\nRSC21",
  "Shen21 CP" = "Shen21\n",
  "Shen21 Mod" = "Mod\nShen21",
  "Shen21 NVV" = "NVX\nShen21",
  "Supasa21 CP LiveVirus" = "Supasa21\n",
  "Supasa21 BNT LiveVirus" = "BNT\nSupasa21",
  "Supasa21 AZD LiveVirus" = "AZD\nSupasa21",
  "Zhou21b CP LiveVirus" = "Zhou21b\n",
  "Zhou21b BNT LiveVirus" = "BNT\nZhou21b",
  "Zhou21b AZD LiveVirus" = "AZD\nZhou21b",
  "Dejnirattisai21 CP LiveVirus" = "Dej21\n",
  "Dejnirattisai21 BNT LiveVirus" = "BNT\nDej21",
  "Dejnirattisai21 AZD LiveVirus" = "AZD\nDej21",
  "Stamatatos21 CP" = "Statmatatos21\n",
  "Stamatatos21 BNT" = "BNT\nSTA21",
  "Stamatatos21 Mod" = "Mod\nSTA21",
  "Tada21 CP" = "Tada21\n",
  "Tada21 BNT" = "BNT\nTada21",
  "Tang21 CP" = "Tang21\n",
  "Tang21 IVIG" = "IVIG\nTang21",
  "Tang21 IVIG LiveVirus" = "IVIG\nTang21*",
  "Wang21 BNT" = "BNT\nWang21",
  "Wang21 Mod" = "Mod\nWang21",
  "Wang21 CP_1M" = "1M\nWang21",
  "Wang21 CP_6M" = "6M\nWang21",
  "Wang21b CP" = "Wang21b\n",
  "Wang21b CP LiveVirus" = "Wang21b*\n",
  "Wang21b BNT" = "BNT\nWang21b",
  "Wang21b BNT LiveVirus" = "BNT\nWang21b*",
  "Wang21b Mod" = "Mod\nWang21b",
  "Wang21b Mod LiveVirus" = "Mod\nWang21b*",
  "Wang21c CP" = "Wang21c\n",
  "Wang21c BNT" = "BNT\nWang21c",
  "Wang21c Mod" = "Mod\nWang21c",
  "Wang21f CP" = "Wang21f\n",
  "Wang21g CP" = "Wang21f\n",
  "Wang21g BBIBP-CorV" = "BBBIBP\nWang21g",
  "Wang21g CoronaVac" = "CoronaVac\nWang21g",
  "Wibmer21 low titer" = "Titer < 400\nWibmer21",
  "Wibmer21 high titer" = "Titer > 400\nWibmer21",
  "Wu21 Mod" = "Mod\nWu21",
  "Widera21 CP LiveVirus" = "Widera21*\n",
  "Widera21 BNT LiveVirus" = "BNT\nWidera21*",
  "Xie21 BNT" = "BNT\nXie21",
  "Xie21b BNT" = "BNT\nXie21b",
  "Skelly21 BNT LiveVirus" = "BNT\nSkelly21",
  "Skelly21 CP" = "Skelly21\n",
  "Planas21 CP_M9" = "Planas21\n",
  "Planas21 BNT_W6" = "BNT\nPlanas21",
  "McCallum21b BNT" = "BNT\nMC21b",
  "McCallum21b CP" = "MC21b\n",
  "McCallum21b MOD LiveVirus" = "MOD\nMC21b"
)
yAxisStep = 10
longRefRxNames = names(shortRefRxNames)

resistance_levels = c(
  'resistant',
  'partial-resistance',
  'susceptible'
)

dfForPlot = dfSuscResultsPlasma %>%
  filter(variant_name %in% filterVariants) %>%
  mutate(
    rx_name = ifelse(is.na(cumulative_group), rx_name, cumulative_group),
    vaccine_name = ifelse(is.na(vaccine_name), "NA", vaccine_name)
  ) %>%
  group_by(ref_name, rx_name, vaccine_name, variant_name, resistance_level, group, assay) %>%
  dplyr::summarise(
    cumulative_count = sum(cumulative_count)
  ) %>%
  mutate(
    cumulative_count = ifelse(is.na(cumulative_count), 0, cumulative_count),
    resistance_level = factor(resistance_level, levels = resistance_levels),
    group = factor(group, levels = c("CP", "IP")),
    ref_rx = sprintf(ifelse(is.na(assay), "%s %s", "%s %s LiveVirus"), ref_name, rx_name)
  ) %>%
  filter(
    ref_rx %notin% excludedRefRxNames
  ) %>%
  mutate(
    ref_rx = factor(ref_rx, levels = unique(c(longRefRxNames, ref_rx)))
  )

plots = list()
i = 1

for (variant in filterVariants) {
  df = dfForPlot %>%
    filter(variant_name == variant)
  totalCount = df %>%
    group_by(ref_rx, group) %>%
    dplyr::summarise(
      total_count = sum(cumulative_count)
    )
  maxCount = ceiling((max(totalCount$total_count)) / yAxisStep) * yAxisStep

  for (groupName in c("CP", "IP")) {
    dfGroup = df %>%
      filter(group == groupName)
    dfVariantLabel = data.frame(
      x = 0,
      y = maxCount,
      text = variantTitle[[variant]]
    )
    numItems = length(unique(dfGroup$ref_rx))
    plot = dfGroup %>%
      ggplot(aes(x = ref_rx, y = cumulative_count)) +
      geom_bar(
        aes(fill = resistance_level),
        size = 0.2,
        color = "#000000",
        stat = 'identity', position = 'stack', width = 0.04 * numItems) +
      scale_x_discrete(
        labels = Vectorize(function(x) ifelse(
          x %in% longRefRxNames,
          yes = shortRefRxNames[[x]],
          no = x
        ))
      ) +
      geom_text(
        aes(x = ref_rx, y = total_count, label = total_count),
        data = totalCount %>% filter(group == groupName),
        size = 3,
        vjust = -0.5
      ) +
      scale_y_continuous(
        expand = c(0, 0, 0.3, 0),
        limits = c(0, 50),
        breaks = seq(0, 50, yAxisStep)
      ) +
      # labs(title = variant) +
      scale_fill_manual(values = c(
        'resistant' = "#146aa8",
        'partial-resistance' = "#7fcbee",
        'susceptible' = "#ffffff"
      )) +
      theme(
        plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size=5),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.placement = "outside",
        panel.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "none",
        panel.border = element_rect(fill = NA, color = 'black')
      )
    if (i < 3) {
      plot = plot + ggtitle(groupTitle[[groupName]])
    }
    if (i %% 2 == 1) {
      plot = plot + annotate(
          "text",
          y=Inf,
          x=0,
          label=paddingTitle(variantTitle[[variant]]),
          size=3,
          vjust=1.5,
          hjust=0
          )
      # plot = plot + geom_text(
      #   aes(x, y, label = text), data = dfVariantLabel,
      #  # hjust = -0.15,
      #  # vjust = -1
      #)
    }
    else {
      plot = plot + theme(
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()
      )
    }
    plots[[i]] = plot
    i = i + 1
  }
}
grid.arrange(grobs = plots, nrow = length(variantTitle))
```

## Monoclonal antibodies

```{r mab-fold-fig, fig.height=20, fig.retina=4, fig.width=24, message=FALSE, warning=FALSE}

groupTitle = c(
  "In clinical trial",
  "Other"
)
variantTitle = c(
  "S:501Y" = "N501Y",
  "B.1.1.7 Spike" = "B.1.1.7",
  "S:484K" = "E484K",
  "B.1.351 Spike" = "B.1.351",
  "P.1 Spike" = "P.1",
  "B.1.429 Spike" = "B.1.427/9",
  "S:452R" = "L452R",
  # "CAL.20C" = "CAL.20C",
  "S:417N" = "K417N",
  "S:439K" = "N439K",
  "S:453F" = "Y453F"
)
filterVariants = names(variantTitle)
filterMAbs = structuredMAbs$ab_name %>%
  setdiff(
    c(
      "DH1041",  # PDB:7LAA, pending
      "5-24",  # PDB:7L2F, pending
      "910-30", # PDB:7KS9, published, pending for structure
      "CV05-163" # PDB:7LOP, pending
    )
  ) %>%
  union(inTrialMAbs$ab_name)

# order of mAbs:
# - MAbs in clinical trial first; add the new trial MAb at the end of this group; well-known ones don't need class/target
# - MAbs with structure data should be ordered by their class: RBM I, 2, 3, RBD-core I, 2
# - Within a mAb class, order mAbs in alphabetical order
mAbLabels = c(
  "Casirivimab" = "CAS\n(RBM I)",
  "Etesevimab" = "ETE\n(RBM I)",
  "Tixagevimab" = "TIX\n(RBM I)",
  "B38" = "B38\n(RBM I)",
  "BD-629" = "BD-629\n(RBM I)",
  "CC12.1" = "CC12.1\n(RBM I)",
  "COVA2-04" = "COVA2-04\n(RBM I)",
  "CV30" = "CV30\n(RBM I)",
  "DH1047" = "DH1047\n(RBM I)",
  "P2C-1F11" = "P2C-1F11\n(RBM I)",
  "S2E12" = "S2E12\n(RBM I)",
  "S2H14" = "S2H14\n(RBM I)",
  "Bamlanivimab" = "BAM\n(RBM II)",
  "Cilgavimab" = "CIL\n(RBM II)",
  "BD-368-2" = "BD-368-2\n(RBM II)",
  "C119" = "C119\n(RBM II)",
  "C121" = "C121\n(RBM II)",
  "P2B-2F6" = "P2B-2F6\n(RBM II)",
  "2-4" = "2-4\n(RBM III)",
  "BD23" = "BD23\n(RBM III)",
  "C144" = "C144\n(RBM III)",
  "S2M11" = "S2M11\n(RBM III)",
  "Imdevimab" = "IMD\n(RBD-core I)",
  "Vir-7831" = "SOT\n(RBD-core I)",
  "C135" = "C135\n(RBD-core I)",
  "COVA1-16" = "COVA1-16\n(RBD-core II)",
  "H014" = "H014\n(RBD-core II)",
  "S2X35" = "S2X35\n(RBD-core II)",
  "Casirivimab+\nImdevimab" = "CAS-IMD",
  "Tixagevimab+\nCilgavimab" = "TIX-CIL",
  "Bamlanivimab+\nEtesevimab" = "BAM-ETE",
  "BRII-196" = "BRII-196",
  "BRII-198" = "BRII-198",
  "BRII-196+\nBRII-198" = "BRII-196+\nBRII-198",
  "JMB2002" = "JMB2002\n(RBD)",
  "4A8" = "4A8\n(NTD)"
)
mAbHasLabels = names(mAbLabels)

dfForPlot = dfSuscResultsMAb %>%
  filter(
    Vectorize(function(x) length(setdiff(x, filterMAbs)) == 0)(ab_name) &
    variant_name %in% filterVariants
  )  %>%
  mutate(
    ref_name = ifelse(
      ref_name == "Rees-Spear21",
      "RSC21",
      ref_name
    ),
    variant_name = factor(variant_name, levels = filterVariants),
    fold_cmp = ifelse(fold > 100, ">", fold_cmp),
    fold = ifelse(fold > 100, 100, fold),
    in_trial = Vectorize(function(x) length(setdiff(x, inTrialMAbs$ab_name)) == 0)(ab_name),
    ab_name = Vectorize(function(x) paste(x, collapse = "+\n"))(ab_name)
  ) %>%
  mutate(
    ab_name = factor(
      ab_name,
      levels = unique(c(mAbHasLabels, ab_name))
    ),
    ref_name = factor(
      ref_name,
      levels = sort(unique(ref_name))
    ),
    resistance_level = factor(
      resistance_level,
      levels = sort(resistance_levels)
    )
  ) %>%
  select(ref_name, ab_name, variant_name, in_trial, resistance_level, fold_cmp, fold) %>%
  group_by(ref_name, ab_name, variant_name, in_trial, fold_cmp) %>%
  dplyr::summarise(
    fold = max(fold),
    resistance_level = tail(sort(resistance_level), n=1)
  )

dfNumAbs = dfForPlot %>%
  group_by(variant_name, in_trial) %>%
  dplyr::summarize(num_abs = length(unique(ab_name)))
dfNumRefs = dfForPlot %>%
  group_by(variant_name, in_trial, ab_name) %>%
  dplyr::summarize(num_refs = length(unique(ref_name))) %>%
  group_by(variant_name, in_trial) %>%
  dplyr::summarize(num_refs = max(num_refs))

dfForPlot = dfForPlot %>%
  inner_join(dfNumAbs, by = c("variant_name", "in_trial")) %>%
  inner_join(dfNumRefs, by = c("variant_name", "in_trial"))

maxNumAbs = max(dfForPlot$num_abs)
maxNumRefs = max(dfForPlot$num_refs)
widthTimer = 0.95 / maxNumAbs / maxNumRefs

plots = list()
i = 1

dfHLines = data.frame(
  y = c(1, 3, 10, 30, 100)
)

for (variant in filterVariants) {
  dfPartial = dfForPlot %>% filter(variant_name == variant) %>% arrange(ref_name)
  dfVariantLabel = data.frame(
    x = 0,
    y = 190,
    variant_name = variant,
    text = variantTitle[[variant]]
  )

  for (cond in c(TRUE, FALSE)) {
    dfPartial2 = dfPartial %>%
      filter(in_trial == cond)
    isEmpty = nrow(dfPartial2) == 0
    if (isEmpty) {
      dfPartial2 = data.frame(
        ref_name = "",
        ab_name = "\n",
        variant_name = variant,
        in_trial = cond,
        fold_cmp = "=",
        fold = 0,
        num_abs = 1,
        num_refs = 1,
        resistance_level = 'susceptible'
      )
    }
    numAbs = dfPartial2$num_abs[[1]]
    numRefs = max(dfPartial2$num_refs)
    plot = dfPartial2 %>%
      ggplot(aes(x = ab_name, y = fold)) +
      geom_hline(aes(yintercept = y), data = dfHLines, linetype = "dashed", color = "#e0e0e0") +
      geom_bar(
        aes(color = ref_name, fill = resistance_level, width = widthTimer * num_abs * num_refs),
        size = 0.2,
        color = "#000000",
        stat = 'identity',
        position = position_dodge2(preserve = 'single')
      ) +
      labs(fill = "References") +
      scale_color_manual(values = rep("#000000", maxNumRefs)) +
      scale_fill_manual(values = c(
        'resistant' = "#146aa8",
        'partial-resistance' = "#7fcbee",
        'susceptible' = "#ffffff"
      )) +
      geom_text(
        aes(
          y = Vectorize(function(x) min(x, 190))(fold),
          label = ifelse(
            fold_cmp == ">",
            sprintf(">%s", formatC(fold, digits = 1, format = 'fg')),
            ""
          )
        ),
        position = position_dodge2(preserve = 'single', width = widthTimer * numAbs * numRefs),
        size = 1.5,
        vjust = -0.5
      ) +
      scale_x_discrete(labels = Vectorize(function(x) ifelse(
        x %in% mAbHasLabels,
        yes = mAbLabels[[x]],
        no = x
      ))) +
      scale_y_continuous(
        trans=scales::pseudo_log_trans(base = 10),
        expand = c(0, 0, 0.1, 0),
        breaks = c(0, 1, 3, 10, 30, 100)
      ) +
      coord_cartesian(ylim = c(0, 200)) +
      theme(
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text = element_blank(),
        panel.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_rect(fill = NA, color = 'black')
      )
    if (cond) {
      plot = plot +
        geom_text(
          aes(x, y, label = text),
          #nudge_x = 0.5,
          data = dfVariantLabel,
          hjust = - 0.2,
          vjust = 0
        )
    }
    else {
      plot = plot +
        theme(
          axis.ticks.y = element_blank(),
          axis.text.y = element_blank()
        )
    }
    if (isEmpty) {
      plot = plot + theme(axis.ticks.x = element_blank())
    }
    if (i %in% c(1, 2)) {
      plot = plot + ggtitle(groupTitle[[i]])
    }
    plots[[i]] = plot
    i = i + 1
  }
}
grid.arrange(grobs = plots, nrow = length(filterVariants))
```
